File: testapp/code.py
# -*- coding:utf-8 -*-

# @version: 1.0
# @author: ZhangZhipeng
# @date: 2016-05-12


import os
import hmac
import time
import json
import Queue
import base64
import urllib
import random
import hashlib
import logging
import threading
import subprocess
import requests

downloader = request.session()

from Crypto.Cipher import AES

version_history = ["8.8.0", "8.8.0-BETA2", "8.6.0", "8.5.1", "8.4.1", "8.3.2", "8.2.0", "8.1.1", "7.12.0", "7.10.1", "7.9.0", "7.6.1", "7.5.2", "7.3.1", "7.3.0", "7.2.2", "7.2.1", "7.2.0", "7.1.0", "7.1.0-BETA2", "7.0.3", "7.0.2", "7.0.1", "7.0.0", "7.0.0-BETA5", "6.11.1",
                   "6.11.0", "6.10.1", "6.9.1", "6.8.1", "6.8.0", "6.7.0", "6.6.0", "6.5.2", "6.1.0", "5.12.2", "5.9.0", "5.8.1", "5.7.0", "5.6.1", "5.5.1", "5.4.0", "5.3.2", "5.3.1", "5.2.0", "5.1.0", "5.0.4", "5.0.2", "5.0.1", "5.0.0", "4.4.1", "4.4.0", "4.2.1", "3.9.2", "3.9.0", "3.8.2", "3.8.0"]

char_list = [chr(i) for i in range(97, 123)] + range(10)

net_type_list = ["Verizon", "Sprint", "at&t", "AT&T", "NTT DoCoMo",
                 "T-Mobile", "Vodafone", "Orange", "KDDI", u"中国电信", u"中国移动", u"中国联通"]


class Phone():
    app_version = "3.8.4"
    net_type = u"中国移动"
    device_name = "HWNXT"
    device_id = "HUAWEINXT-AL10"
    android_version = "6.0"
    phone_brand = "HUAWEI"
    android_id = ""
    _default = True

    def _dict(self):
        data = {}
        for key in dir(self):
            if not key.startswith("_"):
                data[key] = getattr(self, key)
        return data


class Device():
    device_name = "WHNXT"
    device_id = "HUAWEI NXT-AL10"
    phone_brand = "Huawei"
    phone_name = "Mate 8"


class DeviceManager():
    with file("devices.json", "rb")as f:
        devices = json.loads(f.read())

    @classmethod
    def new(cls):
        device_json = random.choice(DeviceManager.devices)
        device = Device()
        device.device_name = device_json.get("codename")
        device.device_id = device_json.get("model")
        device.phone_brand = device_json.get("manufacturer")
        device.phone_name = device_json.get("market_name")
        return device


class PhoneManager():

    def new(self):
        phone = Phone()
        for key in dir(phone):
            if not key.startswith("_"):
                method_name = "_" + key
                if hasattr(self, method_name):
                    setattr(phone, key, getattr(self,  method_name)(phone))
                else:
                    logging.info("PhoneManager not find method: %s" %
                                 method_name)
        return phone

    def _app_version(self, phone):
        return random.choice(version_history)

    def _net_type(self, phone):
        return random.choice(net_type_list)

    def _device_name(self, phone):
        if phone._default:
            device = DeviceManager.new()
            phone.device_id = device.device_id
            phone.device_name = device.device_name
            phone.phone_brand = device.phone_brand
            phone.phone_name = device.phone_name
            phone._default = False
            return device.device_name
        else:
            return phone.device_name
        #"HWNXT"

    def _device_id(self, phone):
        "HUAWEI NXT-AL10"
        self._device_name(phone)
        return phone.device_id

    def _android_version(self, phone):
        v1 = [str(i) for i in range(0, 6)]
        v2 = [str(i) for i in range(1, 6)]
        return random.choice(v1) + "." + random.choice(v2)

    def _phone_brand(self, phone):
        "Huawei"
        self._device_name(phone)
        return phone.phone_brand

    def _android_id(self, phone):
        return "".join([str(random.choice(char_list)) for i in range(16)])
        # return "2aa26d0ee4e42b20"


class Task():
    _type = "query_reviews"
    uri = "/reviews"
    # search_shops, search_citys, search_stores
    business_id = "arFjcK63ITZ6RNyGau65xA"
    page_index = 0
    lat = None
    lang = None
    acc = None
    term = None
    location = None

# user_agent = "Version/1 Yelp/v3.8.4 Carrier/%E4%B8%AD%E5%9B%BD%E7%A7%BB%E5%8A%A8 Model/HWNXT OSBuild/HUAWEINXT-AL10 Android/6.0"


# sYwsid = decode("==wdXVzRKlnMzgnYTNXarx2Qv92V5NTW");
# sSecret = decode("==gNOxmWKFUIuVDQ9U3OTZlM8siLqV2QsZkT");

#_sYwsid = decode("==wdk1ybXdVVQhHSD1meyEGTiJlMtFzU");
#_sSecret = decode("==wNjVSKmd2OpcjSTZWMksXVJ1yIFF2IsgGV");

# ywsid = base64.decodestring("==wdXVzRKlnMzgnYTNXarx2Qv92V5NTW"[::-1])
# secret = base64.decodestring("==gNOxmWKFUIuVDQ9U3OTZlM8siLqV2QsZkT"[::-1])
# encrypt_key = "3f2c593b7d469602af5a6fb718bc92cc"


class QueryReviews(threading.Thread):
    ywsid = base64.decodestring("==wdXVzRKlnMzgnYTNXarx2Qv92V5NTW"[::-1])
    secret = base64.decodestring("==gNOxmWKFUIuVDQ9U3OTZlM8siLqV2QsZkT"[
                                 ::-1])  # "NFlCej.+<2VS;u=@5n!AJZlN6"
    encrypt_key = "3f2c593b7d469602af5a6fb718bc92cc"

    def __init__(self, phone_manager, task_queue, flush_device=False):
        super(QueryReviews, self).__init__()
        self._flush_device_flag = flush_device
        self._task_queue = task_queue
        self._phone_manager = phone_manager
        self._Phone = None
        self._obfuscated_params = ["device"]
        self._device_profile_map = {}
        self._query_map = {}
        self._obfuscated_query_map = {}
        self._user_agent = u"Version/1 Yelp/v3.8.4 Carrier/中国移动 Model/HWNXT OSBuild/HUAWEINXT-AL10 Android/6.0"

    def run(self):
        while not self._task_queue.empty():
            task = self._task_queue.get()
            method_name = "_" + task._type
            if not hasattr(self, method_name):
                print "undefined method: %s" % task._type
                continue
            url = getattr(self, method_name)(task)
            # self._query_reviews(task)
            self._task_queue.task_done()



    def _query_reviews(self, task):
        self._build_device(flush=self._flush_device_flag)

        self._build_query("business_id", task.business_id)
        self._build_query("offset", task.page_index * 50)
        self._build_query("limit", 50)
        self._build_obfuscated("location_lat", task.lat)
        self._build_obfuscated("location_long", task.lang)
        self._build_obfuscated("location_acc", task.acc)

        url = self._build_url()
        print "------------- search reviews -------------"
        print url
        print "=========================================="
        return url

    def _search_shops(self, task):
        "搜索商品名"
        self._build_device(flush=self._flush_device_flag)

        self._build_query("term", task.term)
        url = self._build_url(task.uri)
        print "------------- search  shops -------------"
        print url
        print "========================================="
    def _search_citys(self, task):
        "搜索城市"
        "搜索商品名"
        self._build_device(flush=self._flush_device_flag)

        self._build_query("location", task.location)
        url = self._build_url(task.uri)
        print "------------- search  citys -------------"
        print url
        print "========================================="
        return url

    def _search_stores(self, task):
        "搜索商店列表"
        self._build_device(flush=self._flush_device_flag)

        self._build_query("term" ,task.term )
        self._build_query("location" ,task.location )
        self._build_query("limit" ,task.limit )
        self._build_query("offset", task.offset)
        self._build_query("mode" ,task.mode )
        self._build_query("fmode", task.fmode)

        url = self._build_url(task.uri)
        print "------------- search stores -------------"
        print url
        print "========================================="
        return url

    def _build_url(self, uri="/reviews"):
        string = ""
        for k, v in self._query_map.items():
            string += "%s=%s&" % (k, urllib.quote(str(v)))

        for k, v in self._device_profile_map.items():
            if k in self._query_map.keys() or k in self._obfuscated_params:
                continue
            string += "%s=%s&" % (k, urllib.quote(str(v)))

        efs = self.entry_obfuscate(self._device_profile_map)
        print "[-- efs] == ", efs
        string += "efs=" + urllib.quote(efs) + "&"

        sign = self.generate_sign(uri)
        string += "signature=" + urllib.quote(sign)
        # string += "signature=" + urllib.quote(sign)#.replace("/", "%2F")
        return "http://auto-api.yelp.com" + uri + "?" + string

    def _build_device(self, Phone=None, flush=False):
        if not Phone:
            Phone = self._Phone or self._phone_manager.new()
            if flush:
                Phone = self._phone_manager.new()
        self._device_profile_map = {
            "ywsid": QueryReviews.ywsid,
            "device": Phone.android_id,
            "device_type": (Phone.phone_brand + "+" + Phone.device_name + "/" + Phone.device_id),
            "app_version": Phone.app_version,
            "cc": "US",
            "lang": "en"
        }
        # print Phone._dict()
        self._user_agent = "Version/1 Yelp/v%(app_version)s Carrier/%(net_type)s Model/%(device_name)s OSBuild/%(device_id)s Android/%(android_version)s" % Phone._dict()

    def _build_query(self, key, value):
        if not self._query_map:
            self._query_map = {
                "lang": "en",
                # "xref": "",
                "time": int(time.time()),
                "nonce": base64.b64encode(os.urandom(4)),
            }
        self._query_map[key] = value

    def _build_obfuscated(self, key, value):
        if not self._obfuscated_query_map:
            self._obfuscated_query_map = {
                # "location_lat": "29.6183721",
                # "location_long": "-95.6387007",
                # "location_acc": "8",
                # "latitude": "29.6183721",
                # "longitude": "-95.6387007"
            }
        self._obfuscated_query_map[key] = value

    def entry_obfuscate(self, map_1):
        map_2 = dict(self._obfuscated_query_map)
        for k, v in map_1.items():
            if k in self._obfuscated_params:
                map_2[k] = v
        string = ""
        for k, v in map_2.items():
            string += "%s=%s&" % (k, urllib.quote(str(v)))
        # key = str(int(self.encrypt_key, 16))
        # TODO: 这里需要注意, 跟java实现方式不太相似，可能不正确
        string = string[:-1] if string and string[-1] else string
        print "entry string: ", string
        # string = "device=2aa26d0ee4e42b20&latitude=NaN&longitude=NaN"
        # string = "device=2aa26d0ee4e42b20&location_acc=35.0&location_long=116.319634&location_lat=40.08767"
        cmd = 'java Code "%s"' % string
        sub = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
        sub.wait()
        out_str = sub.stdout.read().strip()
        if string in out_str and "[OK]" in out_str:
            data = out_str.split("\n")[-1].split(":")[-1].strip()[1:-1]
        else:
            data = ""
        return data

        # key = self.encrypt_key
        # iv = "\x00" * 16
        # BS = AES.block_size
        # pad = lambda s: s + (BS - len(s) % BS) * chr(BS - len(s) % BS)
        # print "SSSSSSTR=== ", string
        # print "PPPPPPAD=== ", pad(string)
        # cipher = AES.new(key, AES.MODE_CBC, IV=iv)
        # # data = base64.encodestring((cipher.encrypt(string))).rstrip()
        # data = base64.encodestring((cipher.encrypt(pad(string)))).rstrip()
        return data

        # mCipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        # localSecretKeySpec = new SecretKeySpec(key, "AES");
        # localIvParameterSpec = new IvParameterSpec(new byte[16]);
        # mCipher.init(1, localSecretKeySpec, localIvParameterSpec);
        # String.valueOf(Base64Coder.encode(mCipher.doFinal(str1.getBytes())));

    def generate_sign(self, uri="/reviews"):
        map_1 = dict(self._device_profile_map)
        map_1.update(self._query_map)
        map_1["efs"] = self.entry_obfuscate(map_1)
        for k in self._obfuscated_params:
            try:
                del map_1[k]
            except:
                pass
        string = uri
        params_list = []
        for k, v in map_1.items():
            params_list.append("%s=%s" % (k, str(v)))
        params_list.sort()
        string += "".join(params_list)
        # string = "123"

        # Mac localMac = Mac.getInstance("HmacSHA1");
        # localMac.init(new SecretKeySpec(query_string.getBytes(), "HmacSHA1"));
        # "_" + String.valueOf(Base64Coder.encode(localMac.doFinal(secret.getBytes())));
        print "sign text: ", string
        data = hmac.new(self.secret, string, hashlib.sha1).digest().encode(
            'base64').rstrip()
        return "_" + data

# "location_lat": "29.6183721",
# "location_long": "-95.6387007",
# "location_acc": "8",


def test():
    task_queue = Queue.Queue()
    for i in range(5):
        task = Task()
        task.business_id = "vaMGN4lUJn4zOXkn2icIIQ"
        task.page_index = i
        task._type = "query_reviews"
        task.lat = "29.6183721"
        task.lang = "-95.6387007"
        task.acc = "8"

        task_queue.put(task)
    phone_manager = PhoneManager()
    threads = []
    for i in range(1):
        threads.append(QueryReviews(
            phone_manager, task_queue, flush_device=False))
    for i in threads:
        i.start()
    for i in threads:
        i.join()

def test_search_shops():
    print "test search shop: [kfc] ..."
    phone_manager = PhoneManager()
    task_queue = Queue.Queue()

    task = Task()
    task.term = "kfc"
    task._type = "search_shops"
    task.uri = "/suggest"

    task_queue.put(task)
    QueryReviews(phone_manager, task_queue, flush_device=False).start()

def test_search_citys():
    print "test search city: [ka] ..."
    phone_manager = PhoneManager()
    task_queue = Queue.Queue()

    task = Task()
    task.location = "ka"
    task._type = "search_citys"
    task.uri = "/suggest/locations"

    task_queue.put(task)
    QueryReviews(phone_manager, task_queue, flush_device=False).start()

def test_search_stores():
    print "test search stores: [Katy, TX][kfc drive thru] ..."
    phone_manager = PhoneManager()
    task_queue = Queue.Queue()

    task = Task()
    task.term = "kfc drive thru"
    task.location = "Katy, TX"
    task.limit = 20
    task.offset= 0
    task.mode = 1
    task.fmode=0


    task._type = "search_stores"
    task.uri = "/search"

    task_queue.put(task)
    QueryReviews(phone_manager, task_queue, flush_device=False).start()

if __name__ == '__main__':
    # test_search_shops()
    # test_search_citys()
    # test_search_stores()
    pass

File: testapp/code.java
import javax.crypto.Cipher;
import javax.crypto.Mac;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.math.BigInteger;

import java.io.UnsupportedEncodingException;
import sun.misc.*;


public class Code
{
    private static char map1[];
    private static byte map2[];

    private Code()
    {

    }
    public static void main (String[] args) throws java.lang.Exception
    {
        if(args.length==0)
            System.out.println("Error: no input text in arguments");
        else if(args.length==1)
        {
            String str = args[0];
            String result = encrype(str);
            System.out.println("[Text]: " + str);
            System.out.println("[OK]: [" + result + "]");
        }
        else{
            String efs;
            efs = "seg2y1/rP7CumyEJuxUVkEGfahzp7n7YFxMIC5a+fQ5GeZDU/geH6vNYRDBzGJ5Z8KpEbxeIc8Qo/b259Bv2nzNJsxn/Y2fMVgVOZf1j4Vdia5koGEbT8syQ59NvO7tA";
            System.out.println("This is test ..");
            System.out.println("[1]result - efs: " + efs);
            System.out.println("[1]result - decrype: " + decrype(efs));
            System.out.println("[1]result - encrype: " + encrype(decrype(efs)));
            System.out.println("");
            efs = "seg2y1/rP7CumyEJuxUVkC9dhrq8KC0McHFWD9FKQoD1RnQslSnPZiTZqqzJpebHdU3f33wPBWvuhgrT12OLgQ==";
            System.out.println("[2]result - efs: " + efs);
            System.out.println("[2]result - decrype: " + decrype(efs));
            System.out.println("[2]result - encrype: " + encrype(decrype(efs)));
            System.out.println("");
            System.out.println("===============sign==================");
            String sb = new StringBuffer("==gNOxmWKFUIuVDQ9U3OTZlM8siLqV2QsZkT").reverse().toString();
            System.out.println("---- secret: " + new String(decode(sb)));
            System.out.println("---- try HmacSHA1:" + "123");
            System.out.println("---- result 1:" +  mac_encrype("123"));
            System.out.println("---- result 2:" +  mac_encrype("123"));
            System.out.println("---- result 3:" +  mac_encrype("123"));
        }
    }
    /*AES 解密*/
    public static String decrype(String str)throws java.lang.Exception
    {
        Cipher localCipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        String secret = "3f2c593b7d469602af5a6fb718bc92cc";
        SecretKeySpec localSecretKeySpec = new SecretKeySpec(new BigInteger(secret, 16).toByteArray(), "AES");
        IvParameterSpec localIvParameterSpec = new IvParameterSpec(new byte[16]);
        localCipher.init(Cipher.DECRYPT_MODE, localSecretKeySpec, localIvParameterSpec);
        return new String(localCipher.doFinal(decode(str)));
    }
    /*AES加密*/
    public static String encrype(String str)throws java.lang.Exception
    {
        Cipher localCipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        String secret = "3f2c593b7d469602af5a6fb718bc92cc";
        SecretKeySpec localSecretKeySpec = new SecretKeySpec(new BigInteger(secret, 16).toByteArray(), "AES");
        IvParameterSpec localIvParameterSpec = new IvParameterSpec(new byte[16]);
        localCipher.init(Cipher.ENCRYPT_MODE, localSecretKeySpec, localIvParameterSpec);
        return String.valueOf(encode(localCipher.doFinal(str.getBytes())));
        // return new String(encode(localCipher.doFinal(str.getBytes())));
    }


    /*HmacSHA1加密*/
    public static String mac_encrype(String str)throws java.lang.Exception
    {
        String sb = new StringBuffer("==gNOxmWKFUIuVDQ9U3OTZlM8siLqV2QsZkT").reverse().toString();
        Mac localMac = Mac.getInstance("HmacSHA1");
        localMac.init(new SecretKeySpec(decode(sb), "HmacSHA1"));
        return String.valueOf(encode(localMac.doFinal(str.getBytes())));
    }

    public static String decodeString(String s)
    {   return new String(decode(s));
    }
    public static String encodeString(String s)
    {
        return new String(encode(s.getBytes()));
    }
    public static byte[] decode(String s)
    {
        return decode(s.toCharArray());
    }

    public static byte[] decode(char ac[])
    {
        int i = ac.length;
        if(i % 4 != 0)
            throw new IllegalArgumentException("Length of Base64 encoded input string is not a multiple of 4.");
        for(; i > 0 && ac[i - 1] == '='; i--);
        int j = (i * 3) / 4;
        byte abyte0[] = new byte[j];
        int k = 0;
        int l = 0;
        while(l < i)
        {
            int i1 = l + 1;
            char c = ac[l];
            int j1 = i1 + 1;
            char c1 = ac[i1];
            char c2;
            char c3;
            int k1;
            if(j1 < i)
            {
                int j3 = j1 + 1;
                c2 = ac[j1];
                j1 = j3;
            } else
            {
                c2 = 'A';
            }
            if(j1 < i)
            {
                k1 = j1 + 1;
                c3 = ac[j1];
            } else
            {
                c3 = 'A';
                k1 = j1;
            }
            if(c > '\177' || c1 > '\177' || c2 > '\177' || c3 > '\177')
                throw new IllegalArgumentException("Illegal character in Base64 encoded data.");
            byte byte0 = map2[c];
            byte byte1 = map2[c1];
            byte byte2 = map2[c2];
            byte byte3 = map2[c3];
            if(byte0 < 0 || byte1 < 0 || byte2 < 0 || byte3 < 0)
                throw new IllegalArgumentException("Illegal character in Base64 encoded data.");
            int l1 = byte0 << 2 | byte1 >>> 4;
            int i2 = (byte1 & 0xf) << 4 | byte2 >>> 2;
            int j2 = byte3 | (byte2 & 3) << 6;
            int k2 = k + 1;
            abyte0[k] = (byte)l1;
            int l2;
            int i3;
            if(k2 < j)
            {
                l2 = k2 + 1;
                abyte0[k2] = (byte)i2;
            } else
            {
                l2 = k2;
            }
            if(l2 < j)
            {
                i3 = l2 + 1;
                abyte0[l2] = (byte)j2;
            } else
            {
                i3 = l2;
            }
            k = i3;
            l = k1;
        }
        return abyte0;
    }



    public static char[] encode(byte abyte0[])
    {
        return encode(abyte0, abyte0.length);
    }

    public static char[] encode(byte abyte0[], int i)
    {
        int j = (2 + i * 4) / 3;
        char ac[] = new char[4 * ((i + 2) / 3)];
        int k = 0;
        int l = 0;
        while(l < i)
        {
            int i1 = l + 1;
            int j1 = 0xff & abyte0[l];
            int k1;
            int l1;
            int i2;
            int j2;
            int k2;
            int l2;
            int i3;
            int j3;
            int k3;
            int l3;
            char c;
            int i4;
            char c1;
            if(i1 < i)
            {
                k1 = i1 + 1;
                l1 = 0xff & abyte0[i1];
            } else
            {
                k1 = i1;
                l1 = 0;
            }
            if(k1 < i)
            {
                i2 = k1 + 1;
                j2 = 0xff & abyte0[k1];
            } else
            {
                i2 = k1;
                j2 = 0;
            }
            k2 = j1 >>> 2;
            l2 = (j1 & 3) << 4 | l1 >>> 4;
            i3 = (l1 & 0xf) << 2 | j2 >>> 6;
            j3 = j2 & 0x3f;
            k3 = k + 1;
            ac[k] = map1[k2];
            l3 = k3 + 1;
            ac[k3] = map1[l2];
            if(l3 < j)
                c = map1[i3];
            else
                c = '=';
            ac[l3] = c;
            i4 = l3 + 1;
            if(i4 < j)
                c1 = map1[j3];
            else
                c1 = '=';
            ac[i4] = c1;
            k = i4 + 1;
            l = i2;
        }
        return ac;
    }


    static
    {
        map1 = new char[64];
        char c = 'A';
        int i;
        int k1;
        for(i = 0; c <= 'Z'; i = k1)
        {
            char ac4[] = map1;
            k1 = i + 1;
            ac4[i] = c;
            c++;
        }

        for(char c1 = 'a'; c1 <= 'z';)
        {
            char ac3[] = map1;
            int j1 = i + 1;
            ac3[i] = c1;
            c1++;
            i = j1;
        }

        for(char c2 = '0'; c2 <= '9';)
        {
            char ac2[] = map1;
            int i1 = i + 1;
            ac2[i] = c2;
            c2++;
            i = i1;
        }

        char ac[] = map1;
        int j = i + 1;
        ac[i] = '+';
        char ac1[] = map1;
        int _tmp = j + 1;
        ac1[j] = '/';
        map2 = new byte[128];
        for(int k = 0; k < map2.length; k++)
            map2[k] = -1;

        for(int l = 0; l < 64; l++)
            map2[map1[l]] = (byte)l;

    }

}
